framework/master-key-provider-keyring.md

```diff
@@ -0,0 +1,125 @@
+[//]: # (Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.)
+[//]: # (SPDX-License-Identifier: CC-BY-SA-4.0)
+
+# Master Key Provider Keyring
+
+## Version
+
+0.1.0-preview
+
+## Implementations
+
+- [Java](https://github.com/aws/aws-encryption-sdk-java/blob/master/src/main/java/com/amazonaws/encryptionsdk/keyrings/MasterKeyProviderKeyring.java)
+- [Python](https://github.com/aws/aws-encryption-sdk-python/src/aws_encryption_sdk/keyrings/master_key.py)
+
+## Overview
+
+A keyring which wraps a legacy [Master Key Provider](master-key-provider-interface.md) to facilitate transitioning to [Keyrings](./keyring-interface.md).
+
+### Legacy
+
+Master Key Providers are a legacy construct that have been supplanted by [Keyrings](./keyring-interface.md). 
+The Master Key Provider Keyring should only be used when interfacing with Master Key Providers that cannot otherwise be
+transitioned to keyrings.
+
+## Definitions
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Inputs
+
+On keyring initialization, a keyring MUST define the following:
+
+- [Master Key Provider](#master-key-provider)
+
+### Master Key Provider
+
+A [master key provider](master-key-provider-interface.md) that can decrypt data keys using available [master keys](master-key-interface.md)
+ as well as provide [master keys](master-key-interface.md) for use in encryption.
+
+## Operation
+
+### OnEncrypt
+
+This keyring MUST acquire a list of [master keys](master-key-interface.md) by calling the master key provider's [Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption)
+using the [encryption context](structures.md#encryption-context) from the [encryption materials](structures.md#encryption-materials) as input. If the 
+master key provider does not provide any master keys, OnEncrypt MUST fail.
+
+If the input [encryption materials](structures.md#encryption-materials) do not contain a [plaintext data key](structures.md#plaintext-data-key), OnEncrypt MUST
+attempt to generate a new plaintext data key and encrypt that data key by calling [Generate Data Key](master-key-interface.md#generate-data-key)
+on the first (primary) [master key](master-key-interface.md) provided by the master key provider.
```

Comment:
Again, leave this to the MKP spec.

From MKP#Get_Master_Keys_For_Encryption:
> The output of this interface MUST indicate which one of those master keys MUST be used to generate the data key.

Reply:
Ah ok, the Java ESDK doesn't do that